presto-tests/src/test/java/io/prestosql/execution/TestCoordinatorDynamicFiltering.java

```diff
@@ -0,0 +1,301 @@
+                }
+            });
+        }
+    }
+
+    private static class TestConnector
+            implements Connector
+    {
+        private final ConnectorMetadata metadata;
+        private final Duration scanDuration;
+        private final AtomicReference<TupleDomain<ColumnHandle>> expectedDynamicFilter;
+
+        private TestConnector(ConnectorMetadata metadata, Duration scanDuration, AtomicReference<TupleDomain<ColumnHandle>> expectedDynamicFilter)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.scanDuration = requireNonNull(scanDuration, "scanDuration is null");
+            this.expectedDynamicFilter = requireNonNull(expectedDynamicFilter, "expectedDynamicFilter is null");
+        }
+
+        @Override
+        public ConnectorTransactionHandle beginTransaction(IsolationLevel isolationLevel, boolean readOnly)
+        {
+            return TestingTransactionHandle.create();
+        }
+
+        @Override
+        public ConnectorMetadata getMetadata(ConnectorTransactionHandle transactionHandle)
+        {
+            return metadata;
+        }
+
+        @Override
+        public ConnectorSplitManager getSplitManager()
+        {
+            return new ConnectorSplitManager()
+            {
+                @Override
+                public ConnectorSplitSource getSplits(
+                        ConnectorTransactionHandle transaction,
+                        ConnectorSession session,
+                        ConnectorTableHandle table,
+                        SplitSchedulingStrategy splitSchedulingStrategy,
+                        Supplier<TupleDomain<ColumnHandle>> dynamicFilter)
+                {
+                    long start = System.nanoTime();
+                    return new ConnectorSplitSource()
+                    {
+                        @Override
+                        public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+                        {
+                            return completedFuture(new ConnectorSplitBatch(
+                                    nCopies(maxSize, new EmptySplit(new CatalogName("test"))),
+                                    isFinished()));
+                        }
+
+                        @Override
+                        public void close()
+                        {
+                        }
+
+                        @Override
+                        public boolean isFinished()
+                        {
+                            if (dynamicFilter.get().equals(expectedDynamicFilter.get())) {
+                                // if we received expected dynamic filter then we are done
+                                return true;
+                            }
+                            else if (Duration.nanosSince(start).compareTo(scanDuration) > 0) {
```

Comment:
nanos might flip in comparison, see: https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()

Either use difference for comparison or https://guava.dev/releases/18.0/api/docs/com/google/common/base/Stopwatch.html

Reply:
I think this should ok since `Duration. nanosSince` performs subtraction between the nano time recordings and creates a duration interval from it. I see multiple examples of this way of comparing else where in the code
E.g.: https://github.com/prestosql/presto/blob/f66f62e2cdc752cd30d0acb2040a9a7ec1bde3c0/presto-main/src/main/java/io/prestosql/failuredetector/HeartbeatFailureDetector.java#L358

Reply:
Ah, Right. Thanks!